### PR TITLE
Improve compatibility checks and experience for unsupported versions of PHP

### LINKED ISSUE #5069: Improve compatibility checks and experience for unsupported versions of PHP
## Feature Description

We have become aware that the plugin currently doesn't support each possible combination of PHP & WordPress. We don't currently support below 5.6, however the experience for those in that group is quite poor (as detailed below) and should be improved

It is worth noting that PHP usage below 5.6 is quite low:

5.4 -> 2.7%

@swissspidy to rewrite this exactly how he likes..

PHP 5.2

PHP 5.3

PHP 5.4

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary
Improve compatibility checks and experience for unsupported versions of PHP. 
## Relevant Technical Choices

- Move compatibility class to global namespace and manually load the class file. 
- Move all logic out of namespace.php. 
- Covert Web_Stories_Compability to be PHP 5.2.
- Covert web-stories.php to be PHP 5.2.
- Run WP and WordPress checks on admin notice and WP CLI load. 

## To-do

- Check PHP linter check for PHP 5.2 compat for Web_Stories_Compability.php and web-stories.php. 

## Testing Instructions

---

Fixes #5069

### CODE REVIEW

**Comment by spacedmonkey** [comment]
Screenshot 2020-11-10 at 14 45 04
This is what the notice looks like in WordPress 5.1 + PHP 5.3.

File: web-stories.php [hunk 1]
```diff
@@ -60,15 +65,141 @@
 	define( 'WEBSTORIES_DEV_MODE', false );
 }
 
+// Load Compatibility class the old fashioned way.
+if ( ! class_exists( 'Web_Stories_Compatibility' ) ) {
```
Comment (swissspidy) [inline web-stories.php]: No need for this check? How would it ever already exist?

File: phpcs.xml.dist [hunk 2]
```diff
@@ -58,6 +58,7 @@
 	<!-- The main plugin file should be parsable by PHP < 5.6. -->
 	<rule ref="Generic.Arrays.DisallowLongArraySyntax.Found">
 		<exclude-pattern>web-stories.php</exclude-pattern>
+		<exclude-pattern>Web_Stories_Compatibility.php</exclude-pattern>
```
Comment (swissspidy) [inline phpcs.xml.dist]: This class should not live in the root folder. let's move it to some `compat` folder or something.

Note: any new folder or file in the root should be added to `.distignore`

File: web-stories.php [hunk 3]
```diff
@@ -60,15 +65,141 @@
...
+	require_once WEBSTORIES_PLUGIN_DIR_PATH . 'Web_Stories_Compatibility.php';
+}
+
+global $web_stories_compatibility;
```
Comment (swissspidy) [inline web-stories.php]: We should do `unset( $web_stories_compatibility )` at the end.

File: web-stories.php [hunk 4]
```diff
@@ -60,15 +65,141 @@
...
+);
+
+/**
+ * Displays an admin notice about why the plugin is unable to load.
+ *
+ * @since 1.0.0
+ *
+ * @return void
+ */
+function web_stories_print_admin_notice() {
```
Comment (swissspidy) [inline web-stories.php:144]: Now wondering... Could these functions simply be methods of the compatibility class?

Comment (spacedmonkey) [reply web-stories.php:144]: that doesnt seem right. The compatibility class feels like just the checks. These output are implementations of these checks. Maybe another class...

File: web-stories.php (hunk 1 above)
Comment (spacedmonkey) [reply web-stories.php]: Just being overly careful here.

File: phpcs.xml.dist (hunk 2 above)
Comment (spacedmonkey) [reply phpcs.xml.dist]: Fixed in https://github.com/google/web-stories-wp/pull/5212/commits/79feede30971484f1f72f7c2489605dd29401eed

File: web-stories.php (hunk 3 above)
Comment (spacedmonkey) [reply web-stories.php]: Fixed in https://github.com/google/web-stories-wp/pull/5212/commits/79feede30971484f1f72f7c2489605dd29401eed

**Review by swissspidy — APPROVED** [review]

**Comment by spacedmonkey** [comment]
Screenshot 2020-11-11 at 18 34 32
I tweaked the order of the checks.

File: web-stories.php [hunk 5]
```diff
@@ -60,15 +65,155 @@
...
+/**
+ * Setup web stories compatibility class.
+ *
+ * @SuppressWarnings(PHPMD.MissingImport)
+ *
+ * @since 1.2.0
+ *
+ * @return Web_Stories_Compatibility
+ */
+function get_web_stories_compat_instance() {
```
Comment (swissspidy) [inline web-stories.php]: Nit: Rename to `web_stories_get_compat_instance`

Comment (spacedmonkey) [reply web-stories.php]: Fixed in https://github.com/google/web-stories-wp/pull/5212/commits/41a4bc1a6af21dac139840117c411d0abb6b2c3c

**Pull request merged**